aptos-move/parallel-executor/src/executor.rs

```diff
@@ -45,25 +63,42 @@ impl<'a, K: ModulePath + PartialOrd + Send + Clone + Hash + Eq, V: Send + Sync>
     }
 
     /// Captures a read from the VM execution.
-    pub fn read(&self, key: &K) -> Option<Arc<V>> {
+    pub fn read(&self, key: &K) -> Read<V> {
+        use MVError::*;
+        use Output::*;
+
         loop {
             match self.versioned_map.read(key, self.txn_idx) {
-                Ok((version, v)) => {
+                Ok(Version(version, v)) => {
                     let (txn_idx, incarnation) = version;
-                    self.captured_reads.lock().push(ReadDescriptor::from(
-                        key.clone(),
-                        txn_idx,
-                        incarnation,
-                    ));
-                    return Some(v);
+                    self.captured_reads
+                        .lock()
+                        .push(ReadDescriptor::from_version(
+                            key.clone(),
+                            txn_idx,
+                            incarnation,
+                        ));
+                    return Read::Value(v);
+                }
+                Ok(Resolved(value)) => {
+                    self.captured_reads
+                        .lock()
+                        .push(ReadDescriptor::from_resolved(key.clone(), value));
+                    return Read::U128(value);
                 }
-                Err(None) => {
+                Err(NotFound) => {
                     self.captured_reads
                         .lock()
                         .push(ReadDescriptor::from_storage(key.clone()));
-                    return None;
+                    return Read::None;
+                }
+                Err(Unresolved(delta)) => {
+                    self.captured_reads
+                        .lock()
+                        .push(ReadDescriptor::from_unresolved(key.clone(), delta));
```

Comment:
We have an ABA here. I am convinced it is OK, but it is a bit scary.

Reply:
I think this is not much different from resolved case?

Reply:
Yes, both scary...